README.md
service.subtitles.tusubtitulo
==============================
TuSubtitulo Unofficial subtitle service plugin for XBMC

[ADDON DEAD] - Out of maintenance


changelog.txt
1.0.10
- Fixed link take
1.0.9
- Fixed link take
1.0.8
- Fixed search on the site
1.0.7
- Change way to search subtitles
1.0.6
- Added new parameter for search
1.0.5
- Fixed search on the site
- Fixed download subtitles
1.0.4
- Fixed search on the site
- Fixed download subtitles
1.0.3
- Fixed proxy configuration
1.0.2
- Added alternative name list of series
- Added proxy configuration
- Added some translates
- Clean names subtitles
1.0.1
- Change images without logo of TuSubtitulo
- Added unofficial plugin concept
1.0.0
- Port from Subtitulos.es XBMC subtitles plugin by m0r3
- Fixed recursive expresion to find the url link of subtitles
- Change images with logo of TuSubtitulo

resources/lib/TuSubUtilities.py
# -*- coding: utf-8 -*-

# based on subtitulos.es subtitles
# developed by El_Happy for use TuSubtitulo (unofficially) and XBMC.org

import xbmc
import xbmcaddon
import re
import urllib
from operator import itemgetter
from utils import languages, alternatives
from bs4 import BeautifulSoup

__scriptid__ = xbmcaddon.Addon().getAddonInfo('id')
settings = xbmcaddon.Addon(id=__scriptid__)

main_url = "http://www.tusubtitulo.com/"

def log(module, msg):
  xbmc.log((u"### [%s] - %s" % (module,msg)).encode('utf-8'), level=xbmc.LOGDEBUG)

def search_tvshow(tvshow, season, episode, languages, filename):
  subs = list()
  temp_tvshow = []
  for level in range(5):
    searchstring, ttvshow, sseason, eepisode = getsearchstring(tvshow, season, episode, level)
    log("SEARCHSTRING", searchstring)
    if ttvshow not in temp_tvshow:
      url = main_url + searchstring.lower()
      subs.extend(getallsubsforurl(url, languages, None, ttvshow, sseason, eepisode, level))
      temp_tvshow.append(ttvshow)

  subs = clean_subtitles_list(subs)
  subs = order_subtitles_list(subs)
  return subs

def getsearchstring(tvshow, season, episode, level):

  # Clean tv show name
  if level == 1 and re.search(r'\([^)][a-zA-Z]*\)', tvshow):
    # Series name like "Shameless (US)" -> "Shameless US"
    tvshow = tvshow.replace('(', '').replace(')', '')

  if level == 2 and re.search(r'\([^)][0-9]*\)', tvshow):
    # Series name like "Scandal (2012)" -> "Scandal"
    tvshow = re.sub(r'\s\([^)]*\)', '', tvshow)

  if level == 3 and re.search(r'\([^)]*\)', tvshow):
    # Series name like "Shameless (*)" -> "Shameless"
    tvshow = re.sub(r'\s\([^)]*\)', '', tvshow)

  if level == 4 and tvshow in alternatives:
    # Clean name like "Serie (*)" -> "Serie"
    tvshow = re.sub(r'\s\([^)]*\)', '', tvshow)
    # Search alternative name "Serie" -> "Serie 2014"
    tvshow = alternatives[tvshow]

  # Build search string
  searchstring = 'serie/' + tvshow + '/' + season + '/' + episode + '/*'

  # Replace spaces with dashes
  searchstring = re.sub(r'\s', '-', searchstring)

  # Replace simple comma
  searchstring = re.sub(r'\'', '%27', searchstring)

  return searchstring, tvshow, season, episode

def getallsubsforurl(url, langs, file_original_path, tvshow, season, episode, level):

  subtitles_list = []

  content = geturl(url)
  soup = BeautifulSoup(content, 'html.parser')

  # Search list of subtitles
  for f in soup.find_all(text=re.compile(u'Versión')):
    filename = f.strip()
    if filename != None:
      filename = re.sub(u'Versión', '', filename.strip())
      filename = re.sub(r' ', '.', filename)
      filename = re.sub(r'\s', '.', tvshow) + "." + season + "x" + episode + filename
      filename = re.sub(r'..0.00.megabytes', '', filename)
      filename = re.sub(r'.0.00.megabytes', '', filename)
      filename = re.sub(r', 0.00 megabytes', '', filename)
      server = filename
      backup = filename
      log("FILENAME", filename)

      # Search content of every subtitle
      lang = None
      state = None
      link = None
      for l in f.parent.parent.parent.parent.find_all('li'):
        
        # Take language of subtitle
        if l.find('b') != None and lang == None:
          lang = re.sub(u'ñ', 'n', l.find('b').text.strip())
          lang = re.sub(u'á', 'a', lang)
          lang = re.sub(u'é', 'e', lang)
          lang = re.sub(u'à', 'a', lang)
          log("LANG", lang)

          if lang in languages:
            languageshort = languages[lang][1]
            languagelong = languages[lang][0]
            filename = filename + ".(%s)" % languages[lang][2]
            server = filename
            order = 1 + languages[lang][3]
          else:
            lang = "Unknown"
            languageshort = languages[lang][1]
            languagelong = languages[lang][0]
            filename = filename + ".(%s)" % languages[lang][2]
            server = filename
            order = 1 + languages[lang][3]
          continue

        # Take state of subtitle
        if l.text.strip() == 'Completado' and state == None:
          state = l.text.strip()
          log("STATE", state)
          continue

        # Take link of subtitle
        if l.find('a', href=re.compile(r'\w+\/\d+\/\d+')) != None and link == None:
          link = main_url + l.find('a', href=re.compile(r'\w+\/\d+\/\d+')).get('href')
          log("LINK", link)

        # Just add complete subtitles
        if lang != None and state != None and link != None and languageshort in langs:
          subtitles_list.append({'no_files': 1, 'filename': filename, 'server': server, 'sync': False, 'language_flag': languageshort + '.gif', 'language_name': languagelong, 'hearing_imp': False, 'link': link, 'lang': languageshort, 'order': order})
          lang = None
          state = None
          link = None
          filename = backup
          server = backup

  return subtitles_list

def geturl(url):
  class AppURLopener(urllib.FancyURLopener):
    version = "User-Agent=Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36"
    def __init__(self, *args):
      urllib.FancyURLopener.__init__(self, *args)
    def add_referrer(self, url=None):
      if url:
        urllib._urlopener.addheader('Referer', url)

  if settings.getSetting('PROXY') == 'true':
    log("PROXY", "Proxy enabled")
    proxy = {settings.getSetting('PROXY_PROTOCOL') : settings.getSetting('PROXY_PROTOCOL') + '://' + settings.getSetting('PROXY_HOST') + ':' + settings.getSetting('PROXY_PORT')}
    urllib._urlopener = AppURLopener(proxy)
  else:
    log("PROXY", "Proxy disabled")
    urllib._urlopener = AppURLopener()
  urllib._urlopener.add_referrer(main_url)
  try:
    response = urllib._urlopener.open(url)
    content = response.read()
  except:
    content = None
  return content

def clean_subtitles_list(subtitles_list):
  seen = set()
  subs = []
  for sub in subtitles_list:
    filename = sub['link']
    if filename not in seen:
      subs.append(sub)
      seen.add(filename)
  return subs

def order_subtitles_list(subtitles_list):
  return sorted(subtitles_list, key=itemgetter('order'))

service.py
# -*- coding: utf-8 -*- 

"""
This addon is a modification of m0r3's addon for Subtitulos.es, adapted to TuSubtitulo

Original code by m0re (https://github.com/m0r3)
"""

import os
import sys
import xbmc
import urllib
import xbmcvfs
import xbmcaddon
import xbmcgui
import xbmcplugin
import re
import shutil
import unicodedata

__addon__ = xbmcaddon.Addon()
__scriptid__   = __addon__.getAddonInfo('id')
settings = xbmcaddon.Addon(id=__scriptid__)

__cwd__        = xbmc.translatePath( __addon__.getAddonInfo('path') ).decode("utf-8")
__profile__    = xbmc.translatePath( __addon__.getAddonInfo('profile') ).decode("utf-8")
__resource__   = xbmc.translatePath( os.path.join( __cwd__, 'resources', 'lib' ) ).decode("utf-8")
__temp__       = xbmc.translatePath( os.path.join( __profile__, 'temp') ).decode("utf-8")

sys.path.append (__resource__)

from TuSubUtilities import search_tvshow, log

""" Called when searching for subtitles from XBMC """
def Search(item):
  subs = search_tvshow(item['tvshow'], item['season'], item['episode'], item['2let_language'], item['file_original_path'])
  for sub in subs:
    append_subtitle(sub)

def append_subtitle(item):
  listitem = xbmcgui.ListItem(label=item['language_name'], label2=item['filename'], thumbnailImage=item['lang'])

  listitem.setProperty("sync",  'true' if item["sync"] else 'false')
  listitem.setProperty("hearing_imp", 'true' if item["hearing_imp"] else 'false')

  ## below arguments are optional, it can be used to pass any info needed in download function
  ## anything after "action=download&" will be sent to addon once user clicks listed subtitle to downlaod
  url = "plugin://%s/?action=download&link=%s&filename=%s" % (__scriptid__, item['link'], item['filename'])

  ## add it to list, this can be done as many times as needed for all subtitles found
  xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]), url=url, listitem=listitem, isFolder=False)

""" Called when subtitle download request from XBMC """
def Download(link, filename):
  subtitle_list = []
  exts = [".srt", ".sub", ".txt", ".smi", ".ssa", ".ass"]

  if link:
    log(__name__, "Downloadlink %s" % link)

    class MyOpener(urllib.FancyURLopener):
      version = "User-Agent=Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36"
      def __init__(self, *args):
        urllib.FancyURLopener.__init__(self, *args)

    if settings.getSetting('PROXY') == 'true':
      proxy = {settings.getSetting('PROXY_PROTOCOL') : settings.getSetting('PROXY_PROTOCOL') + '://' + settings.getSetting('PROXY_HOST') + ':' + settings.getSetting('PROXY_PORT')}
      my_urlopener = MyOpener(proxy)
    else:
      my_urlopener = MyOpener()

    my_urlopener.addheader('Referer', link)
    log(__name__, "Fetching subtitles using url '%s' with referer header '%s'" % (link, link))
    response = my_urlopener.open(link)

    local_tmp_file = os.path.join(__temp__, "sub.srt")
    
    if xbmcvfs.exists(__temp__):
      shutil.rmtree(__temp__)
    xbmcvfs.mkdirs(__temp__)
    try:
      log(__name__, "Saving subtitles to '%s'" % local_tmp_file)
      local_file_handle = open(local_tmp_file, "wb")
      local_file_handle.write(response.read())
      local_file_handle.close()
      
      subtitle_list.append(local_tmp_file)
      log(__name__, "=== returning subtitle file %s" % file)

    except:
      log(__name__, "Failed to save subtitle to %s" % local_tmp_file)

  return subtitle_list

""" Get parameters from XBMC and launch actions """
def normalizeString(str):
  return unicodedata.normalize(
         'NFKD', unicode(unicode(str, 'utf-8'))
         ).encode('ascii','ignore')    
 
def get_params():
  param=[]
  paramstring=sys.argv[2]
  if len(paramstring)>=2:
    params=paramstring
    cleanedparams=params.replace('?','')
    if (params[len(params)-1]=='/'):
      params=params[0:len(params)-2]
    pairsofparams=cleanedparams.split('&')
    param={}
    for i in range(len(pairsofparams)):
      splitparams={}
      splitparams=pairsofparams[i].split('=')
      if (len(splitparams))==2:
        param[splitparams[0]]=splitparams[1]

  return param

params = get_params()

if params['action'] == 'search':
  item = {}
  item['temp']               = False
  item['rar']                = False
  item['year']               = xbmc.getInfoLabel("VideoPlayer.Year")                           # Year
  item['season']             = str(xbmc.getInfoLabel("VideoPlayer.Season"))                    # Season
  item['episode']            = str(xbmc.getInfoLabel("VideoPlayer.Episode"))                   # Episode
  item['tvshow']             = normalizeString(xbmc.getInfoLabel("VideoPlayer.TVshowtitle"))   # Show
  item['title']              = normalizeString(xbmc.getInfoLabel("VideoPlayer.OriginalTitle")) # try to get original title
  item['file_original_path'] = urllib.unquote(xbmc.Player().getPlayingFile().decode('utf-8'))  # Full path of a playing file
  item['3let_language']      = []
  item['2let_language']      = []

  for lang in urllib.unquote(params['languages']).decode('utf-8').split(","):
    item['3let_language'].append(xbmc.convertLanguage(lang,xbmc.ISO_639_2))
    item['2let_language'].append(xbmc.convertLanguage(lang,xbmc.ISO_639_1))

  if item['title'] == "":
    item['title']  = normalizeString(xbmc.getInfoLabel("VideoPlayer.Title"))      # no original title, get just Title

  if item['episode'].lower().find("s") > -1:                                      # Check if season is "Special"
    item['season'] = "0"                                                          #
    item['episode'] = item['episode'][-1:]

  if ( item['file_original_path'].find("http") > -1 ):
    item['temp'] = True

  elif ( item['file_original_path'].find("rar://") > -1 ):
    item['rar']  = True
    item['file_original_path'] = os.path.dirname(item['file_original_path'][6:])

  elif ( item['file_original_path'].find("stack://") > -1 ):
    stackPath = item['file_original_path'].split(" , ")
    item['file_original_path'] = stackPath[0][8:]

  # required if tvshow is not indexed/recognized in library
  if item['tvshow'] == "":  
    log(__name__, "item %s" % item)
    # replace dots with spaces in title
    titulo = re.sub(r'\.', ' ', item['title'])
    log(__name__, "title no dots: %s" % titulo)
    mo = re.search(r'(.*)[sS](\d+)[eE](\d+)', titulo) #S01E02 like
    if not mo:
      mo = re.search(r'(.*)(\d\d)[xX](\d+)', titulo) # old 10x02 style
    if not mo:
      mo = re.search(r'(.*)(\d)[xX](\d+)', titulo) # old 1x02 style
    if not mo:
      mo = re.search(r'(.*) (\d+)(\d\d)', titulo) # 102 style 
    # split title in tvshow, season and episode
    if mo:
      item['tvshow'] = mo.group(1)
      item['season'] = mo.group(2)
      item['episode'] = mo.group(3)
      log(__name__, "item %s" % item)
    else:
      log(__name__, "could not parse tvshow name and episode number")

  Search(item)  

elif params['action'] == 'download':
  ## we pickup all our arguments sent from def Search()
  subs = Download(params["link"],params["filename"])
  ## we can return more than one subtitle for multi CD versions, for now we are still working out how to handle that in XBMC core
  for sub in subs:
    listitem = xbmcgui.ListItem(label=sub)
    xbmcplugin.addDirectoryItem(handle=int(sys.argv[1]),url=sub,listitem=listitem,isFolder=False)

xbmcplugin.endOfDirectory(int(sys.argv[1])) ## send end of directory to XBMC
